src/OpenTelemetry.Exporter.InMemory/README.md

```diff
@@ -17,6 +17,16 @@ See the
 [`TestInMemoryExporter.cs`](../../examples/Console/TestInMemoryExporter.cs) for
 an example of how to use the exporter for exporting traces to a collection.
 
+## Remarks
+
+When working with `Metric` it's important to note that by design the
+MetricApi reuses Metrics (see also: `MetricReader.metricsCurrentBatch`).
+This means that after multiple exports, the `InMemoryExporter` will
```

Comment:
I think this line is not by design, it's a bug that we want to fix (later).

Reply:
Making sure we're talking about the same thing...

- "MetricApi reusing Metric instances" - This is by design, correct?
- We may fix the InMemoryExporter, but after today's meeting we're leaning towards no change. I'm not sure we want to state this as a bug in this doc.

Reply:
Why do we think users of the InMemoryExporter will have to keep in mind / note that the SDK is doing some special optimization? This statement itself sounds like a design flaw to me.

Reply:
@cijothomas @alanwest Are we going to keep InMemoryExporter in beta/prerelease? I agree with @reyang. Before we go stable we should have a solution for this. Some kind of safe copy thing into the collection. Because if not we are taking a dependency on the internal behavior in the public API and that might prevent us from changing the internals in the future?

Reply:
The fact that SDK is re-using Metric for optimization/whatever purpose is fine. But the fact that this detail is "leaking" into exporters is the issue we are discussing.

Several options before us: (In the order of my preference)

1. Do nothing for 1.2 stable release. i.e proceed with a known issue/limitation. 
2. Document that Exporters cannot assume anything about the state of the telemetry it received in the Export(), once control returns from Export() method. 
     The above document will be applicable to **every exporters**, not just InMemoryExporter.
3. Even with above document, there is no "clone" method for the Export authors. We may address this by adding new Clone()/equivalent method. Do we need to block 1.2 for this? I'd say no - this repo itself has used InMemoryExporter for testing, even with the current limitation.
4. Modify the SDK pipeline, to always provide a snapshot/copy of Metric, which is unaffected any further action in the SDK. This requires some re-architecting, and will likely affect performance negatively. 
 
      (Not exactly same, but related issue exists in Traces also, as any other processor/exporter *can* modify the Activity, affecting other exporters. We have not prevented this issue, apart from documenting it. One possible approach we briefly discussed during tracing release was to modify Exporters to receive a ReadOnlyActivity which prevents it from modifying Activity...)

Reply:
@cijothomas How do you feel about adding a new immutable class in the exporter project to add into the collection of exported items instead of `Metric`? I was messing around yesterday with that idea: https://github.com/open-telemetry/opentelemetry-dotnet/compare/main...CodeBlanch:inmemory-metrics-copy?expand=1

Preserves the semantics of the exporter without needing to re-architect the internals.

Reply:
I think this does address the issue for InMemoryExporter. Its a bit of "cheating" by exposing internals specially to this Exporter :smile: 

The statement "exporters cannot assume anything about the state of the telemetry it received in the Export(), once control returns from Export() method." is still applicable for every other exporter, right?

Reply:
It is for sure "cheating" 🤣 But I wouldn't say it is exposing the internals. I think what we have today, capturing `Metric` into collection, is exposing the internals and that is what is causing us pain. What this change does is introduce a safe class for the scenario so we don't have to expose/re-architect the internals 😄

> The statement "exporters cannot assume anything about the state of the telemetry it received in the Export(), once control returns from Export() method." is still applicable for every other exporter, right?

Definitely 💯 and I don't think we should change that behavior. It would be very expensive to always make copies in order to give exporter something safe to use beyond `Export`.

Reply:
ok. 
so every exporter still need abide by the rule that "take whatever you want in the Export() call itself".

InMemoryExporter is used for testing, and it stores Metrics into a user provided collection. InMemoryExporter has the ability currently (once we apply your suggesttion PR), to snap/clone the state when it adds to the collection. This is done to allow writing tests which need to compare the data between multiple exports.

^ is this everyone's understanding?
If no, lets clarify before proceeding further.

Reply:
Just to confirm for next steps
- Short Term (this PR): we want to keep the change to `docs/metrics/extending-the-sdk/README.md` and remove other changes.
- Long Term: do a more thorough investigation of cloning exports

Reply:
Short Term (this PR): we want to keep the change to docs/metrics/extending-the-sdk/README.md and remove other changes.

^ Yes.

Reply:
Define long term 😄 I don't think we should release `OpenTelemetry.Exporter.InMemory` 1.2 (stable) without addressing the API we are exposing there. If we are going to keep it prerelease until then, 👍

Reply:
Yea I now realized that this would be breaking change if done after stable.
So agree, we should do this before 1.2.

Reply:
> Do we need to block 1.2 for this? I'd say no - this repo itself has used InMemoryExporter for testing, even with the current limitation.

^ this comment of mine is incorrect, as fixing it after 1.2 is breaking change for InMemoryExporter.